module/zfs/zfs_vnops.c

```diff
@@ -3885,11 +3901,38 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 			goto out;
 		}
 	}
+	/* Target must exist for RENAME_EXCHANGE. */
+	if (!tzp && txtype == TX_EXCHANGE) {
+		error = SET_ERROR(ENOENT);
+		goto out;
+	}
```

Comment:
nit: what do you think about moving this check in to an `else` block to remove the redundant `tzp` check?

```c

        /*
         * Does target exist?
         */
        if (tzp) {
                ...
        } else {
                /* Target must exist for RENAME_EXCHANGE. */
                if (txtype == TX_EXCHANGE) {
                        error = SET_ERROR(ENOENT);
                        goto out;
                }
        }
```